### PR TITLE
Power iteration Eigenvalue calculation

### PR DESCRIPTION
**Pull request by Beakerboy**
This is another algorithm that easily calculates both eigenvalues and eigenvectors, so It might be worthwhile considering how to combine the two static classes into one.

### CODE REVIEW

**Comment by markrogoyski** [comment]
Thanks for the pull request that provides another good eigenvalue method.

File: src/LinearAlgebra/Eigenvalue.php [hunk 1]
```diff
@@ -89,4 +104,49 @@ public static function closedFormPolynomialRootMethod(Matrix $A): array
...
+        $m = $A->getM();
+        $random_array = [];
+        for ($i = 0; $i <$m; $i++) {
+            $random_array[] = [rand()];
+        }
+        $b = MatrixFactory::create($random_array);
+        $newμ = 0;
+        $μ = -1;
+        $new_abs = -2;
+        while (!Support::isEqual($μ, $newμ)) {
```
Comment (markrogoyski) [inline src/LinearAlgebra/Eigenvalue.php]: Does there need to be some alternate max iterations condition or something just as a check to make sure we don't go into an infinite loop on some edge case?

File: tests/LinearAlgebra/EigenvalueTest.php [hunk 2]
```diff
@@ -20,6 +20,21 @@ public function testClosedFormPolynomialRootMethod(array $A, array $S)
...
+     * @testCase     powerIterationMethod returns the expected eigenvalues
+     * @dataProvider dataProviderForEigenvalues
+     * @param        array $A
+     * @param        array $S
+     */
+    public function testPowerIteration(array $A, array $S)
+    {
+        $A = MatrixFactory::create($A);
+        // Find the most extreme value
+        $eigenvalue = max($S) ** 2 > min($S) ** 2 ? max($S) : min($S);
```
Comment (markrogoyski) [inline tests/LinearAlgebra/EigenvalueTest.php]: We write unit tests to validate our code that is doing complex computations. Generally you don't want the test code itself doing complicated computations which could introduce mistakes into the test. For this test, it is probably best to just add a third parameter to the data provider that represents the greatest absolute value eigenvalue.

I was thinking, maybe this would be a useful function to write, `absoluteMax`. Give it an array of numbers and it finds the value with the greatest absolute value. Would something like that be generally useful? And if so, what would it return for a list like [-4, 1, 3, 4]? Would you get 4 or -4?

Comment (Beakerboy) [reply tests/LinearAlgebra/EigenvalueTest.php]: I was thinking that, but I feel that for an arbitrarily simple expression like this, it makes more sense to have data providers that can be used in multiple tests, then to add in extra fluff that is needed for this one test. As other eigenvalue methods come into the class, any specific test data made for them can all be passed through the powerIteration function.

Comment (markrogoyski) [reply tests/LinearAlgebra/EigenvalueTest.php]: Yeah, I actually started writing that method, and then when I wrote my first couple test cases, I kind of got stuck.

For the data provider, just FYI, your data provider could have 10 parameters, but if you test only uses the first two, your test method declaration only needs to declare the first two. This is fine.

Your data provider could look like this:
```php
    public function dataProviderForEigenvalues(): array
    {
        return [
            [
                [
                    [0, 1],
                    [-2, -3],
                ],
                [-2, -1],
                -2
            ],
        ];
    }
```

And you can have two tests using different elements:
```php
    public function testClosedFormPolynomialRootMethod(array $A, array $S)
    {
        // ...
    }

    public function testPowerIteration(array $A, array $S, int $expectedGreatestAbsoluteEigenvalue)
    {
        // ...
    {
```

Consider if this kind of generic, yet multi-purpose kind of data provider would be useful here.

Comment (Beakerboy) [reply tests/LinearAlgebra/EigenvalueTest.php]: Ahh, that's better than I was assuming, but having to include `array $S` as a pointless function parameter still bugs me. I'll do it.

File: src/LinearAlgebra/Eigenvalue.php (hunk 1 above)
Comment (Beakerboy) [reply src/LinearAlgebra/Eigenvalue.php]: I added a new exception and updated this.

File: src/LinearAlgebra/Eigenvalue.php [hunk 3]
```diff
@@ -89,4 +104,54 @@ public static function closedFormPolynomialRootMethod(Matrix $A): array
...
+     *
+     * will converge to the dominant eigenvector,
+     *
+     * The corresponding eigenvalue is calculated as:
+     *
+     *      bₖᐪAbₖ
+     * μₖ = -------
+     *       bₖᐪbₖ
+     *
+     * https://en.wikipedia.org/wiki/Power_itera
```
Comment (markrogoyski) [inline src/LinearAlgebra/Eigenvalue.php]: Looks like a copy and paste mistake or something on the end of this URL.

File: src/LinearAlgebra/Eigenvalue.php [hunk 4]
```diff
@@ -89,4 +104,54 @@ public static function closedFormPolynomialRootMethod(Matrix $A): array
...
+     * @param Matrix $Ation
+     * @param int $iterations max number of iterations to perform
+     *
+     * @return float[] most extreme eigenvalue
```
Comment (markrogoyski) [inline src/LinearAlgebra/Eigenvalue.php]: Since this only returns a single eigenvalue, would it make more sense to just return the value and not wrap it in an array? The `Matrix::eigenvalues` has to return an array, but this method doesn't seem to need to return an array. However, if this returns just the value, then the `Matrix::eigenvalues` needs to know what to do with each method and how to handle the response to provide an array of eigenvalues per its contract. To simplify this interface, `Matrix::eigenvalues` needs to add some complexity to handle this simplification. What do you think? I tend to lean towards simple public interfaces even at the expense of complexity in the implementations.

File: src/LinearAlgebra/Eigenvalue.php (hunk 3 above)
Comment (Beakerboy) [reply src/LinearAlgebra/Eigenvalue.php]: Fixed

File: src/LinearAlgebra/Eigenvalue.php (hunk 4 above)
Comment (Beakerboy) [reply src/LinearAlgebra/Eigenvalue.php]: My thought was to keep the return type the same as other implementations. Just like how the result of a 1x1 Matrix will be a single value from any other algorithm, but we would still return it as an array.

File: src/LinearAlgebra/Eigenvalue.php [hunk 5]
```diff
@@ -89,4 +104,54 @@ public static function closedFormPolynomialRootMethod(Matrix $A): array
...
+    public static function powerIteration(Matrix $A, int $iterations = 1000): array
+    {
+        self::checkMatrix($A);
+        
+        $m = $A->getM();
+        $random_array = [];
+        for ($i = 0; $i <$m; $i++) {
+            $random_array[] = [rand()];
+        }
+        $b = MatrixFactory::create($random_array);
```
Comment (markrogoyski) [inline src/LinearAlgebra/Eigenvalue.php]: How about replacing all of this code with a new factory method that describes what we want?
```php
$b = MatrixFactory::random($A->getM(), 1);
```

We want a random m x 1 matrix to initialize the power iteration. The work to create this matrix is not relevant to this algorithm and distracts from what is the focus here. So let's move this responsibility somewhere else that cares about creating different kinds of Matrices.

Comment (Beakerboy) [reply src/LinearAlgebra/Eigenvalue.php]: done

File: tests/LinearAlgebra/MatrixFactoryTest.php [hunk 6]
```diff
@@ -837,4 +838,17 @@ public function testHilbertExceptionNLessThanZero()
...
+
+    /**
+     * Test that the appropriate matrix is constructed
+     *
+     * @runInSeparateProcess
+     */
+    public function testRandomMatrix()
+    {
+        $expected = [[.31415926535]];
+        $rand = $this->getFunctionMock('MathPHP\LinearAlgebra', 'rand');
```
Comment (markrogoyski) [inline tests/LinearAlgebra/MatrixFactoryTest.php]: Neat!

File: tests/LinearAlgebra/MatrixFactoryTest.php [hunk 7]
```diff
@@ -837,4 +838,17 @@ public function testHilbertExceptionNLessThanZero()
         $this->expectException(Exception\OutOfBoundsException::class);
         MatrixFactory::hilbert(-1);
     }
+
+    /**
+     * Test that the appropriate matrix is constructed
+     *
+     * @runInSeparateProcess
```
Comment (markrogoyski) [inline tests/LinearAlgebra/MatrixFactoryTest.php]: Is this necessary? I looked through PHPMock and it looks like when you call `getFunctionMock` it registers the FunctionMock as a listener on the test result. At the end of the test, it will call each listeners' `endTest` method, of which the PHPMock's MockDisabler will deactivate the mock. Try it without this annotation and see if it works.

Comment (Beakerboy) [reply tests/LinearAlgebra/MatrixFactoryTest.php]: I had issues in another project without this. If rand() is called before the FunctionMock is registered, calls to rand() will go to \rand() despite the existence of a namspaced function. So if any test in the MathPHP\LinearAlgebra namespace calls rand(), this will be needed. I'll check...

File: composer.json [hunk 8]
```diff
@@ -16,7 +16,8 @@
       "squizlabs/php_codesniffer": "3.*",
       "phpstan/phpstan": "*",
       "phpmd/phpmd": "^2.6",
-      "phploc/phploc": "*"
+      "phploc/phploc": "*",
+      "php-mock/php-mock-phpunit": "*"
```
Comment (markrogoyski) [inline composer.json]: Probably best to be specific with the release. Use 2.4 or 2.*.

Comment (Beakerboy) [reply composer.json]: Got it. I was thinking it defaulted to most recent acceptable version.

**Comment by Beakerboy** [comment]
the @runInSeparateProcess annotation IS needed. Also I've been meaning to ask, the @testCase annotaion is not an official PHPUnit annotation. is there some other purpose for it?

**Comment by markrogoyski** [comment]
> the @runInSeparateProcess annotation IS needed. Also I've been meaning to ask, the @testcase annotaion is not an official PHPUnit annotation. is there some other purpose for it?

It was a mistake. A while ago I updated the coding standards: https://github.com/markrogoyski/math-php/wiki/Coding-Standards#unit-tests

It should have been `@test` and not `testcase`. The reason to add the `@test` annotations is like a safety net. PHPUnit will run any method that starts with the word "test" or has the @test annotation. In the past, multiple times (on this project and others) there were unit tests that were named improperly and never run. So, adding the annotation is like a second chance to get it right, to make sure PHPUnit runs the test.

File: tests/LinearAlgebra/MatrixFactoryTest.php (hunk 7 above)
Comment (markrogoyski) [reply tests/LinearAlgebra/MatrixFactoryTest.php]: OK. Thanks for investigating.

**Comment by markrogoyski** [comment]
Thanks for working on the power iteration eigenvalue method.
